### PR TITLE
qs: implement quasisymmetry as a Python geometry objective

### PR DESCRIPTION
**Pull request by krystophny**
Stack 3/4. Targets `main`; merge after #703.

## Scope

This PR implements SIMSOPT's `QuasisymmetryRatioResidual` as a Python/JAX objective over `vmecpp.geometry`.

The residual includes the flux-surface measure, covariant current averages, iota, and the requested helicity. It contains no C++, pybind, Enzyme, VMEC solver change, or QS-specific native kernel. QS is the demonstration objective for the generic geometry interface.

## Tests

- reconstruct `|B|` and compare it with VMEC's independent `bmnc` spectrum
- compare the residual vector and total with SIMSOPT on a 3D equilibrium
- differentiate the scalar objective with JAX

The SIMSOPT comparison gives a residual-vector cosine above 0.9999 and a total within `1e-3` relative tolerance.

Stack: #584 -> #703 -> **#707** -> #710.

### CODE REVIEW

**Comment by krystophny** [comment]
This is the direct answer to your 23 Aug review, and it is now visible in the diff rather than just asserted.

**#707 is 126 lines of Python and nothing else.** No C++, no pybind, no Enzyme, no QS kernel, no VMEC changes. Quasisymmetry is written purely against the `vmecpp.geometry` contract from #703: `|B|` is reconstructed from the R, Z, lambda and flux jets, and the objective is normalized non-QS Fourier power for QA or a chosen helicity.

That is the property you asked for. A different QI definition, a turbulence proxy, or anything else is written exactly this way, by a researcher, without touching VMEC++ and without us shipping a kernel for it. #710 then supplies the adjoint for free.

Previously this PR contained the autodiff layer instead of the QS layer (the stack was off by one); that is corrected.

Verified: `test_qs.py` passes - zero QA residual on an axisymmetric equilibrium, nonzero and differentiable on a 3D perturbation.

**Comment by krystophny** [comment]
Restacked onto current `main` and the updated #584/#703. Focused local validation passes: geometry + QS tests are 5/5. One Ubuntu 22.04 / Python 3.13 job failed before installation because Google's Chrome apt mirror was mid-sync (`Packages.gz` size mismatch); this is unrelated infrastructure. I do not have permission to rerun upstream jobs, so please rerun that failed job after the matrix completes. Ready for review after #584 and #703 land.

**Comment by jurasic-pf** [comment]
Nit: Can we move it to simsopt compat, or something like that? Loiks great and works great though! 

Example/downstream (shouldn't ship as a first-class core module). it's a demo for the geometry contract and largely overlaps something SIMSOPT already owns (with the difference being vjp capability). It'd fit better under `examples/`

**Comment by jurasic-pf** [comment]
Patch for the move out of the core package, since I cannot push to the fork branch.

`vmecpp/qs.py` becomes the private `vmecpp/_qs.py`, and `magnetic_field_strength`, `quasisymmetry_residuals` and `quasisymmetry_total` are re-exported from `vmecpp.simsopt_compat`, next to the SIMSOPT wrapper whose `QuasisymmetryRatioResidual` the objective mirrors. `simsopt` is already a required dependency of `vmecpp`, so nothing is added to the dependency set.

Local runs on this branch (GCC build): `tests/test_qs.py` 3 passed; `tests/test_geometry.py` plus `tests/test_simsopt_compat.py` 49 passed, 1 skipped.

<details><summary>707: qs -> _qs, re-exported from simsopt_compat</summary>

```diff

diff --git a/src/vmecpp/qs.py b/src/vmecpp/_qs.py
similarity index 98%
rename from src/vmecpp/qs.py
rename to src/vmecpp/_qs.py
index 93385051..574dba2f 100644
--- a/src/vmecpp/qs.py
+++ b/src/vmecpp/_qs.py
@@ -1,5 +1,8 @@
 """Quasisymmetry objectives built only from the public geometry API.
 
+Re-exported from vmecpp.simsopt_compat, next to the SIMSOPT wrapper this
+mirrors; not part of the solver.
+
 The metric is the one SIMSOPT's ``QuasisymmetryRatioResidual`` implements: for
 quasisymmetry the ratio ``(B x grad B . grad psi) / (B . grad B)`` is constant
 on a flux surface, so
diff --git a/src/vmecpp/simsopt_compat.py b/src/vmecpp/simsopt_compat.py
index 9778c831..50baf454 100644
--- a/src/vmecpp/simsopt_compat.py
+++ b/src/vmecpp/simsopt_compat.py
@@ -21,6 +21,13 @@ from vmecpp import (  # noqa: F401
     ensure_vmecpp_input,
     is_vmec2000_input,
 )
+from vmecpp._qs import (  # noqa: F401
+    # Quasisymmetry objectives, the VMEC++ counterpart to
+    # simsopt.mhd.vmec_diagnostics.QuasisymmetryRatioResidual.
+    magnetic_field_strength,
+    quasisymmetry_residuals,
+    quasisymmetry_total,
+)
 
 logger = logging.getLogger(__name__)
 
diff --git a/tests/test_qs.py b/tests/test_qs.py
index acf4c739..ecceaff5 100644
--- a/tests/test_qs.py
+++ b/tests/test_qs.py
@@ -3,7 +3,8 @@ import jax.numpy as jnp
 import numpy as np
 
 import vmecpp
-from vmecpp import geometry, qs
+from vmecpp import geometry
+from vmecpp import simsopt_compat as qs
 from vmecpp.cpp import _vmecpp  # type: ignore
 
 jax.config.update("jax_enable_x64", True)
-- 
2.55.0

```

</details>

#710 needs the same rename followed through. Merging this into `pr707-solve-vjp` is a clean rename merge; the remaining three references are the docstring example in `autodiff.py` and the import in `tests/test_autodiff.py`.

Local run on the merged result (GCC build): 7 passed, 2 skipped, the skips being the two Enzyme-gated cases including `test_quasisymmetry_gradient_through_a_three_dimensional_solve`.

<details><summary>710: follow the rename</summary>

```diff
diff --git a/src/vmecpp/autodiff.py b/src/vmecpp/autodiff.py
index bd151195..966b5c30 100644
--- a/src/vmecpp/autodiff.py
+++ b/src/vmecpp/autodiff.py
@@ -447,10 +447,12 @@ def make_solver(vmec_input) -> DifferentiableVmec:
 
     Example:
 
-        from vmecpp import autodiff
+        from vmecpp import autodiff, simsopt_compat
 
         solver = autodiff.make_solver(input)
-        objective = lambda boundary: qs.quasisymmetry_total(solver(boundary), [0.6])
+        objective = lambda boundary: simsopt_compat.quasisymmetry_total(
+            solver(boundary), [0.6]
+        )
         value, gradient = jax.value_and_grad(objective)(boundary)
 
     Forward execution and the VJP both invoke VMEC++ in memory. The VJP is
diff --git a/tests/test_autodiff.py b/tests/test_autodiff.py
index a1b4684a..ad8e1c16 100644
--- a/tests/test_autodiff.py
+++ b/tests/test_autodiff.py
@@ -7,7 +7,8 @@ import pytest
 from scipy.sparse.linalg import LinearOperator, gmres
 
 import vmecpp
-from vmecpp import autodiff, geometry, qs
+from vmecpp import autodiff, geometry
+from vmecpp import simsopt_compat as qs
 from vmecpp.cpp import _vmecpp  # type: ignore
 
 jax.config.update("jax_enable_x64", True)
-- 
2.55.0

```

</details>

File: tests/test_qs.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
+import jax
+import jax.numpy as jnp
+import numpy as np
+
+import vmecpp
+from vmecpp import geometry, qs
```
Comment (jurasic-pf) [inline tests/test_qs.py]:
```suggestion
from vmecpp import geometry
from vmecpp import simsopt_compat as qs
```

**Review by jurasic-pf — COMMENTED** [review]
Sorry, I thought I could just inline edit your PR with the suggested fix, but GH is somewhat limited in what it allows maintainers to edit on PRs.
I'll force merge partially broken and fix CI myself, just to unblock. You did enough work
